flang/include/flang/Optimizer/Dialect/FIROps.td

```diff
@@ -1113,6 +1113,56 @@ def fir_EmboxOp : fir_Op<"embox", [NoSideEffect, AttrSizedOperandSegments]> {
   }];
 }
 
+def fir_ReboxOp : fir_Op<"rebox", [NoSideEffect, AttrSizedOperandSegments]> {
+  let summary = "create a box given another box and (optional) dimension information";
+
+  let description = [{
+    Create a new boxed reference value from another box. This is meant to be used
+    when the taking a reference to part of a boxed value, or to an entire boxed value with
+    new shape or type information.
+
+    A slice argument must be provided to build a reference to part of a boxed value.
+    In this case, an optional fir.shift argument can be passed if the slice indexes
+    are not one based. This fir.shift will not be used to set the new fir.box lower
+    bounds that will be one if no new lower bounds are provided in newShape operand.
+
+    The new extra information can be:
+      - new shape information (New lower bounds, new rank, or new extents.
+        New rank/extents can only be provided if the original fir.box is
+        contiguous in all dimension but maybe the first one). newShape
+        operand must be provided to set new shape information.
+      - new type (only for derived types). It is possible to set the dynamic type
+        of the new box to one of the parent types of the input box dynamic type.
+        Type parameters cannot be changed. This change is reflected in the requested
+        result type of the new box.
+
+    It is possible to pass both a slice argument and new shape/type arguments. In this
+    case, the new shape/type are applied on the result of the slice. Note that the shape
+    argument is not used to indicate the slice index origins, a dedicated argument, sliceOrigins,
+    is intended for that.
+
+    ```mlir
+      // TODO
+    ```
+
+  }];
+
+  let arguments = (ins
+    fir_BoxType:$box,
+    Optional<AnyShapeOrShiftType>:$newShape,
+    Optional<fir_SliceType>:$slice,
+    Optional<fir_ShiftType>:$sliceOrigins
```

Comment:
Shouldn't this be folded in with the `newShape`? I'm not sure why we'd allow a `shape_shift` for `newShape` and a second `shift` for `sliceOrigins`? BTW, the `slice` itself already contains a set of lower bounds, to orient which subspace is being selected by the slice operation.

Reply:
The new shape applies after the fir.box slicing, so it's lower bounds do not relate to the slice origin. Since sliceOps are not zero based, its origins need an ad-hoc argument to be passed.

Here is a use case:

```fortran
 subroutine bar(p)
  real, pointer :: p(:, :, :)
  ! ...
end subroutine

subroutine foo(x, i, j, k, l, m)
  integer :: i, j, k, l, m
  real, target :: x(i:, j:)
  real, pointer :: p(:, :, :)
  p(k:5, l:7, m:9) => x(10:1000:16, 11)
  call bar(p)
end subroutine
```

To implement this, the `newShape` is a `shape_shift` with lower bounds `k`, `l`, and `m`. But the origins of the slice `(10:1000:16, 11)` are `i` and `j`. Codegen needs to know about `i` and `j` to apply the slice correctly.

The lower bounds that must be given to the resulting entity of the fir.rebox are not the one that applies to the slice. There may actually also be an issue with fir.embox that is making the assumption that the lower bounds of the created fir.box and the ones to be used for the slice are the same.

Reply:
I'd prefer these be entirely independent operations. Slicing is one thing. Applying dimensions another. It should be the case that `fir.rebox` is composable in this way.

Reply:
I have changed the ops and refactored codegen so that rebox can either be used for slicing or reshaping a box but not both at the same time anymore. So my example above would now be implemented in two successive fir.rebox.